python/ql/src/semmle/python/ApiGraphs.qll

```diff
@@ -0,0 +1,417 @@
+        lbl = Label::memberFromRef(ref) and
+        ref = pred.getAnAttributeRead()
+        or
+        // Calling a node that is a use of `base`
+        lbl = Label::return() and
+        ref = pred.getACall()
+      )
+    }
+
+    /**
+     * Holds if `ref` is a use of node `nd`.
+     */
+    cached
+    predicate use(TApiNode nd, DataFlow::Node ref) {
+      exists(string name |
+        nd = MkModuleImport(name) and
+        imports(ref, name)
+      )
+      or
+      nd = MkUse(ref)
+    }
+
+    /**
+     * Gets a data-flow node to which `nd`, which is a use of an API-graph node, flows.
+     *
+     * The flow from `nd` to that node may be inter-procedural.
+     */
+    private DataFlow::LocalSourceNode trackUseNode(
+      DataFlow::LocalSourceNode src, DataFlow::TypeTracker t
+    ) {
+      t.start() and
+      use(_, src) and
+      result = src
+      or
+      // Due to bad performance when using `trackUseNode(t2, attr_name).track(t2, t)`
+      // we have inlined that code and forced a join
+      exists(DataFlow::StepSummary summary |
+        t = trackUseNode_first_join(src, result, summary).append(summary)
+      )
+    }
+
+    pragma[nomagic]
+    private DataFlow::TypeTracker trackUseNode_first_join(
+      DataFlow::LocalSourceNode src, DataFlow::LocalSourceNode res, DataFlow::StepSummary summary
+    ) {
+      DataFlow::StepSummary::step(trackUseNode(src, result), res, summary)
+    }
+
+    cached
+    DataFlow::LocalSourceNode trackUseNode(DataFlow::LocalSourceNode src) {
+      result = trackUseNode(src, DataFlow::TypeTracker::end())
+    }
+
+    /**
+     * Holds if there is an edge from `pred` to `succ` in the API graph that is labeled with `lbl`.
+     */
+    cached
+    predicate edge(Node pred, string lbl, Node succ) {
+      /* There's an edge from the root node for each imported module. */
+      exists(string m |
+        pred = MkRoot() and
+        lbl = Label::mod(m)
+      |
+        succ = MkModuleImport(m) and
+        // Only allow undotted names to count as base modules.
+        not m.matches("%.%")
+      )
+      or
+      /* Step from the dotted module name `foo.bar` to `foo.bar.baz` along an edge labeled `baz` */
+      exists(string member |
+        prefix_member(pred, member, succ) and
+        lbl = Label::member(member)
+      )
+      or
+      /* Every node that is a use of an API component is itself added to the API graph. */
+      exists(DataFlow::LocalSourceNode ref |
+        use(pred, lbl, ref) and
+        succ = MkUse(ref)
+      )
+    }
+
+    /**
+     * Holds if there is an edge from `pred` to `succ` in the API graph.
+     */
+    private predicate edge(TApiNode pred, TApiNode succ) { edge(pred, _, succ) }
+
+    /** Gets the shortest distance from the root to `nd` in the API graph. */
+    cached
+    int distanceFromRoot(TApiNode nd) = shortestDistances(MkRoot/0, edge/2)(_, nd, result)
+  }
+}
+
+private module Label {
```

Comment:
It feels like this should be providing an IPA type rather than a bunch of functions returning strings. But I do not feel too strongly about this..

Reply:
I don't see what we would gain from this, other than having to add extra machinery to translate instances of said IPA type into strings. (Unless you _also_ want to represent paths as an IPA type, in which case life becomes very painful and recursive.)

Reply:
Indeed. Perhaps I just read the code as typed before I got to the module..